### PR TITLE
fleetd: support conflict in global unit

### LINKED ISSUE #1109: Support Conflicts when Global=true
When `Global=true` is set in the `[X-Fleet]` section, the only additional option applied is `MachineMetadata`. The `Conflicts` option should also be supported such that a machine must pass both the `MachineMetadata` and `Conflicts` filters in order to run a given unit. If `MachineMetadata` is not set but `Conflicts` is, only the `Conflicts` option must be checked.

### PR DESCRIPTION
**Pull request by wuqixuan**
In global unit, support Conflict flag. So in
some machine, the unit will not be launched.

Fixes #1109

### CODE REVIEW

File: engine/state.go [hunk 1]
```diff
@@ -93,9 +93,15 @@ func (cs *clusterState) agents() map[string]*agent.AgentState {
...
 		gu := gu
 		for _, a := range agents {
-			if machine.HasMetadata(a.MState, gu.RequiredTargetMetadata()) {
-				a.Units[gu.Name] = gu
+			if !machine.HasMetadata(a.MState, gu.RequiredTargetMetadata()) {
+				continue
 			}
+
+			if cExists, _ := a.HasConflict(gu.Name, gu.Conflicts()); cExists {
+				//fmt.Printf("%s cannot start conflit with %s at %s\n", gu.Name, cJobName, a.MState.ID);
```
Comment (bcwaldon) [inline engine/state.go]: log this at debug

**Comment by bcwaldon** [comment]
@wuqixuan Please update the documentation to reflect this new support

**Comment by wuqixuan** [comment]
@bcwaldon , updated as per comments.

**Comment by wuqixuan** [comment]
@mischief @bcwaldon, do you have another comment for my latest modification?

**Comment by wuqixuan** [comment]
Gently ping @jonboulle

File: fleetctl/fleetctl_test.go [hunk 2]
```diff
@@ -165,32 +165,26 @@ func TestCreateUnitFails(t *testing.T) {
 		{
 			"foo.service",
 			newUnitFile(t, `[X-Fleet]
-	MachineOf=abcd
-	Conflicts=abcd`),
+			MachineOf=abcd
```
Comment (jonboulle) [inline fleetctl/fleetctl_test.go]: why this change?

File: agent/reconcile.go [hunk 3]
```diff
@@ -141,17 +141,36 @@ func desiredAgentState(a *Agent, reg registry.Registry) (*AgentState, error) {
 
 	for _, u := range units {
```
Comment (jonboulle) [inline agent/reconcile.go]: hrm, why break this into two loops and not just have a switch within the loop?

**Comment by jonboulle** [comment]
Any chance of a functional test for this one, too?

**Comment by dongsupark** [comment]
Closed in favor of https://github.com/coreos/fleet/pull/1571
